File: src/incode_mcp/token_batching.py
"""Tokenizer-bounded window planning and microbatch packing.

Character windows bound characters, not the token count that drives embedding
memory. Attention cost is quadratic in sequence length, so the same 4,096
characters cost wildly different amounts depending on how densely they tokenize:
ordinary source is ~984 tokens, a minified line is ~2,157, and embedding the
latter as one sequence adds ~1,172 MiB of resident memory against ~266 MiB for
the same characters split into token-bounded windows.

Everything here is pure and tokenizer-agnostic so the policy is testable without
loading a model. The only tokenizer contact is :func:`content_token_offsets`,
which reads an already-produced encoding.
"""

from __future__ import annotations

from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import Any

DEFAULT_MAX_TOKENS = 1024
DEFAULT_OVERLAP_TOKENS = 64
# Microbatch packing budget: item_count * longest_padded_tokens. Padding is to
# the longest member, so this bounds the padded matrix a batch materializes.
DEFAULT_MAX_TOKEN_PRODUCT = 4096
# A candidate never exceeds the extractor's character ceiling, so a window
# fan-out above this implies a pathological tokenization rather than dense code.
# It is a tripwire, not a working limit: 4,096 characters cannot tokenize to
# more than 4,096 tokens, which is five windows at the default budget.
MAX_WINDOWS_PER_CANDIDATE = 16


@dataclass(frozen=True)
class TokenWindow:
    """A token-bounded slice of one candidate, in candidate-relative characters."""

    start_char: int
    end_char: int
    token_count: int


def content_token_offsets(encoding: Any) -> list[tuple[int, int]]:
    """Return character spans for real tokens, dropping ``[CLS]``/``[SEP]``.

    *encoding* is anything with ``offsets`` and, optionally, a
    ``special_tokens_mask`` — a ``tokenizers.Encoding`` in production. Special
    tokens carry a ``(0, 0)`` span, so leaving them in would make every window
    appear to start at the beginning of the text.
    """
    offsets: list[tuple[int, int]] = [(int(start), int(end)) for start, end in encoding.offsets]
    mask = getattr(encoding, "special_tokens_mask", None)
    if mask is None:
        return offsets
    return [span for span, special in zip(offsets, mask, strict=True) if not special]


def plan_token_windows(
    offsets: Sequence[tuple[int, int]],
    *,
    text_length: int,
    max_tokens: int = DEFAULT_MAX_TOKENS,
    overlap_tokens: int = DEFAULT_OVERLAP_TOKENS,
    max_windows: int = MAX_WINDOWS_PER_CANDIDATE,
) -> list[TokenWindow]:
    """Split a candidate into windows of at most *max_tokens* tokens.

    Windows are contiguous in characters and overlap by *overlap_tokens* tokens,
    so no source is dropped between them. Boundaries depend only on the
    tokenization, never on the memory budget or on how a batch was packed, so a
    retry at a smaller microbatch size re-derives the identical windows.
    """
    if max_tokens < 1:
        raise ValueError("max_tokens must be at least 1")
    if text_length < 0:
        raise ValueError("text_length must not be negative")
    # Half the budget is the widest overlap that keeps the window count within
    # twice the minimum. Without it a budget squeezed by a wide prefix would
    # leave a one-token stride and fan a candidate out into hundreds of windows.
    overlap = min(max(overlap_tokens, 0), max_tokens - 1, max_tokens // 2)
    stride = max_tokens - overlap
    total = len(offsets)
    if total == 0:
        # Whitespace-only or untokenizable content still needs one window so the
        # candidate keeps a vector rather than silently vanishing from the index.
        return [TokenWindow(0, text_length, 0)] if text_length else []
    if total <= max_tokens:
        return [TokenWindow(0, text_length, total)]

    windows: list[TokenWindow] = []
    start_token = 0
    while start_token < total:
        end_token = min(start_token + max_tokens, total)
        # Character bounds run from this token's start to the *next* token's
        # start, so inter-token whitespace stays attached to the earlier window
        # and the concatenated windows cover the candidate exactly.
        start_char = 0 if start_token == 0 else offsets[start_token][0]
        end_char = text_length if end_token == total else offsets[end_token][0]
        windows.append(TokenWindow(start_char, end_char, end_token - start_token))
        if len(windows) > max_windows:
            raise ValueError(
                f"Token planning exceeded {max_windows} windows for a "
                f"{text_length}-character candidate ({total} tokens)"
            )
        if end_token == total:
            break
        start_token += stride
    return windows


def plan_candidate_windows(
    encode: Callable[[str], Any],
    candidates: Sequence[tuple[str, str]],
    *,
    max_tokens: int = DEFAULT_MAX_TOKENS,
    overlap_tokens: int = DEFAULT_OVERLAP_TOKENS,
    max_windows: int = MAX_WINDOWS_PER_CANDIDATE,
) -> list[list[TokenWindow]]:
    """Plan windows for ``(prefix, content)`` candidates.

    The prefix is the context header repeated on every window of a candidate, so
    it is charged against the budget once and the content windows are sized with
    what is left.
    """
    plans: list[list[TokenWindow]] = []
    prefix_tokens: dict[str, int] = {}
    for prefix, content in candidates:
        if prefix not in prefix_tokens:
            prefix_tokens[prefix] = len(content_token_offsets(encode(prefix))) if prefix else 0
        # Keep at least one token of forward progress per window even when a
        # pathological prefix would otherwise consume the whole budget.
        budget = max(overlap_tokens + 1, max_tokens - prefix_tokens[prefix])
        plans.append(
            plan_token_windows(
                content_token_offsets(encode(content)),
                text_length=len(content),
                max_tokens=budget,
                overlap_tokens=overlap_tokens,
                max_windows=max_windows,
            )
        )
    return plans


def plan_microbatches(
    token_counts: Sequence[int],
    *,
    max_items: int = 1,
    max_token_product: int = DEFAULT_MAX_TOKEN_PRODUCT,
) -> list[list[int]]:
    """Bucket segment indices by length, then stay within both packing limits.

    A batch pads to its longest member, so ``item_count * longest`` — not the
    sum — is what the model materializes. A single segment always forms a batch
    even when it exceeds the product on its own; there is nothing smaller to
    fall back to. Power-of-two buckets keep similarly sized segments together
    without making exact token counts part of the ordering contract.
    """
    if max_items < 1:
        raise ValueError("max_items must be at least 1")
    ordered = sorted(
        enumerate(token_counts),
        key=lambda item: (
            -1 if item[1] > max_token_product else max(0, item[1]).bit_length(),
            item[0],
        ),
    )
    batches: list[list[int]] = []
    current: list[int] = []
    longest = 0
    for index, count in ordered:
        widened = max(longest, count)
        if current and (
            len(current) + 1 > max_items or (len(current) + 1) * widened > max_token_product
        ):
            batches.append(current)
            current = []
            widened = count
        current.append(index)
        longest = widened
    if current:
        batches.append(current)
    return batches

File: ts/packages/server/src/token-batching.ts
/**
 * Tokenizer-bounded window planning and microbatch packing.
 *
 * Character windows bound characters, not the token count that drives embedding
 * memory. Attention cost is quadratic in sequence length, so the same 4,096
 * characters cost wildly different amounts depending on how densely they
 * tokenize: ordinary source is ~984 tokens, a minified line is ~2,157, and
 * embedding the latter as one sequence adds ~1,172 MiB of resident memory
 * against ~266 MiB for the same characters split into token-bounded windows.
 *
 * Everything here is pure and tokenizer-agnostic so the policy is testable
 * without loading a model. The only tokenizer contact is `contentTokenOffsets`,
 * which reads an already-produced encoding.
 */

export const DEFAULT_MAX_TOKENS = 1024;
export const DEFAULT_OVERLAP_TOKENS = 64;
/**
 * Microbatch packing budget: itemCount * longestPaddedTokens. Padding is to the
 * longest member, so this bounds the padded matrix a batch materializes.
 */
export const DEFAULT_MAX_TOKEN_PRODUCT = 4096;
/**
 * The indexing ceiling `DEFAULT_MAX_TOKEN_PRODUCT` was measured against. The
 * product bounds the padded matrix a microbatch materializes, so it is a memory
 * budget expressed in tokens; leaving it fixed while the configured ceiling
 * moves either wastes the memory the operator granted or overruns the memory
 * they withheld.
 */
export const REFERENCE_MEMORY_BYTES = 2 * 1024 ** 3;
/**
 * Padding cost is quadratic in the widest member, and nothing has measured a
 * matrix beyond this multiple of the reference. A ceiling larger than that buys
 * no further width rather than extrapolating past the evidence.
 */
export const MAX_TOKEN_PRODUCT_MULTIPLE = 8;
/**
 * A candidate never exceeds the extractor's character ceiling, so a window
 * fan-out above this implies a pathological tokenization rather than dense code.
 * It is a tripwire, not a working limit: 4,096 characters cannot tokenize to
 * more than 4,096 tokens, which is five windows at the default budget.
 */
export const MAX_WINDOWS_PER_CANDIDATE = 16;

/** A token-bounded slice of one candidate, in candidate-relative characters. */
export interface TokenWindow {
  readonly startChar: number;
  readonly endChar: number;
  readonly tokenCount: number;
}

/** A character span, `[start, end)`, as a tokenizer reports it. */
export type TokenSpan = readonly [start: number, end: number];

/**
 * The slice of a tokenizer encoding this module reads.
 *
 * Structural rather than nominal, exactly as the Python duck-typing was: the
 * backend adapter in Phase 4 shapes whatever the bindings hand back into this,
 * and the tests supply a deterministic word-splitter instead.
 */
export interface TokenEncoding {
  readonly offsets: readonly TokenSpan[];
  readonly specialTokensMask?: readonly number[] | null | undefined;
}

/**
 * Return the padded-token budget a *memoryBytes* indexing ceiling supports.
 *
 * The floor is one longest window: a product below that would not admit even a
 * single max-length sequence as a batch of one, and would drop every shorter
 * segment to one item per batch on the way there.
 */
export function maxTokenProductFor(
  memoryBytes: number,
  { maxTokens = DEFAULT_MAX_TOKENS }: { maxTokens?: number } = {},
): number {
  const scaled = Math.floor(
    (DEFAULT_MAX_TOKEN_PRODUCT * Math.max(0, memoryBytes)) / REFERENCE_MEMORY_BYTES,
  );
  const ceiling = DEFAULT_MAX_TOKEN_PRODUCT * MAX_TOKEN_PRODUCT_MULTIPLE;
  return Math.max(maxTokens, Math.min(scaled, ceiling));
}

/**
 * Return character spans for real tokens, dropping `[CLS]`/`[SEP]`.
 *
 * Special tokens carry a `(0, 0)` span, so leaving them in would make every
 * window appear to start at the beginning of the text.
 */
export function contentTokenOffsets(encoding: TokenEncoding): TokenSpan[] {
  const offsets = encoding.offsets.map(([start, end]): TokenSpan => [start, end]);
  const mask = encoding.specialTokensMask;
  if (mask === undefined || mask === null) return offsets;
  if (mask.length !== offsets.length) {
    throw new Error(`special token mask covers ${mask.length} of ${offsets.length} offsets`);
  }
  return offsets.filter((_span, index) => !mask[index]);
}

/**
 * Split a candidate into windows of at most *maxTokens* tokens.
 *
 * Windows are contiguous in characters and overlap by *overlapTokens* tokens, so
 * no source is dropped between them. Boundaries depend only on the tokenization,
 * never on the memory budget or on how a batch was packed, so a retry at a
 * smaller microbatch size re-derives the identical windows.
 */
export function planTokenWindows(
  offsets: readonly TokenSpan[],
  {
    textLength,
    maxTokens = DEFAULT_MAX_TOKENS,
    overlapTokens = DEFAULT_OVERLAP_TOKENS,
    maxWindows = MAX_WINDOWS_PER_CANDIDATE,
  }: {
    textLength: number;
    maxTokens?: number;
    overlapTokens?: number;
    maxWindows?: number;
  },
): TokenWindow[] {
  if (maxTokens < 1) throw new Error("maxTokens must be at least 1");
  if (textLength < 0) throw new Error("textLength must not be negative");
  // Half the budget is the widest overlap that keeps the window count within
  // twice the minimum. Without it a budget squeezed by a wide prefix would leave
  // a one-token stride and fan a candidate out into hundreds of windows.
  const overlap = Math.min(Math.max(overlapTokens, 0), maxTokens - 1, Math.floor(maxTokens / 2));
  const stride = maxTokens - overlap;
  const total = offsets.length;
  if (total === 0) {
    // Whitespace-only or untokenizable content still needs one window so the
    // candidate keeps a vector rather than silently vanishing from the index.
    return textLength ? [{ startChar: 0, endChar: textLength, tokenCount: 0 }] : [];
  }
  if (total <= maxTokens) {
    return [{ startChar: 0, endChar: textLength, tokenCount: total }];
  }

  const windows: TokenWindow[] = [];
  let startToken = 0;
  while (startToken < total) {
    const endToken = Math.min(startToken + maxTokens, total);
    // Character bounds run from this token's start to the *next* token's start,
    // so inter-token whitespace stays attached to the earlier window and the
    // concatenated windows cover the candidate exactly.
    const startChar = startToken === 0 ? 0 : (offsets[startToken]?.[0] ?? 0);
    const endChar = endToken === total ? textLength : (offsets[endToken]?.[0] ?? textLength);
    windows.push({ startChar, endChar, tokenCount: endToken - startToken });
    if (windows.length > maxWindows) {
      throw new Error(
        `Token planning exceeded ${maxWindows} windows for a ` +
          `${textLength}-character candidate (${total} tokens)`,
      );
    }
    if (endToken === total) break;
    startToken += stride;
  }
  return windows;
}

/** One `(prefix, content)` candidate awaiting a window plan. */
export interface WindowCandidate {
  readonly prefix: string;
  readonly content: string;
}

/**
 * Plan windows for `(prefix, content)` candidates.
 *
 * The prefix is the context header repeated on every window of a candidate, so
 * it is charged against the budget once and the content windows are sized with
 * what is left.
 */
export function planCandidateWindows(
  encode: (text: string) => TokenEncoding,
  candidates: readonly WindowCandidate[],
  {
    maxTokens = DEFAULT_MAX_TOKENS,
    overlapTokens = DEFAULT_OVERLAP_TOKENS,
    maxWindows = MAX_WINDOWS_PER_CANDIDATE,
  }: { maxTokens?: number; overlapTokens?: number; maxWindows?: number } = {},
): TokenWindow[][] {
  const plans: TokenWindow[][] = [];
  const prefixTokens = new Map<string, number>();
  for (const { prefix, content } of candidates) {
    let charged = prefixTokens.get(prefix);
    if (charged === undefined) {
      charged = prefix ? contentTokenOffsets(encode(prefix)).length : 0;
      prefixTokens.set(prefix, charged);
    }
    // Keep at least one token of forward progress per window even when a
    // pathological prefix would otherwise consume the whole budget.
    const budget = Math.max(overlapTokens + 1, maxTokens - charged);
    plans.push(
      planTokenWindows(contentTokenOffsets(encode(content)), {
        textLength: content.length,
        maxTokens: budget,
        overlapTokens,
        maxWindows,
      }),
    );
  }
  return plans;
}

/**
 * Bucket segment indices by length, then stay within both packing limits.
 *
 * A batch pads to its longest member, so `itemCount * longest` -- not the sum --
 * is what the model materializes. A single segment always forms a batch even
 * when it exceeds the product on its own; there is nothing smaller to fall back
 * to. Power-of-two buckets keep similarly sized segments together without making
 * exact token counts part of the ordering contract.
 */
export function planMicrobatches(
  tokenCounts: readonly number[],
  {
    maxItems = 1,
    maxTokenProduct = DEFAULT_MAX_TOKEN_PRODUCT,
  }: { maxItems?: number; maxTokenProduct?: number } = {},
): number[][] {
  if (maxItems < 1) throw new Error("maxItems must be at least 1");
  const bucket = (count: number): number =>
    count > maxTokenProduct ? -1 : bitLength(Math.max(0, count));
  const ordered = tokenCounts
    .map((count, index) => ({ count, index }))
    .sort((left, right) => bucket(left.count) - bucket(right.count) || left.index - right.index);

  const batches: number[][] = [];
  let current: number[] = [];
  let longest = 0;
  for (const { count, index } of ordered) {
    let widened = Math.max(longest, count);
    if (current.length > 0) {
      if (current.length + 1 > maxItems || (current.length + 1) * widened > maxTokenProduct) {
        batches.push(current);
        current = [];
        widened = count;
      }
    }
    current.push(index);
    longest = widened;
  }
  if (current.length > 0) batches.push(current);
  return batches;
}

/** `int.bit_length()`: how many bits the magnitude of a non-negative int needs. */
function bitLength(value: number): number {
  return value === 0 ? 0 : 32 - Math.clz32(value);
}
